Style navigation QuickNav command

mltony opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review". 
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
Closes #16000
### Summary of the issue:
Adding jump to same style and jump to different style QuickNav commands.
### Description of user facing changes
Adding jump to same style and jump to different style QuickNav commands. They are not assigned to any keyboard gestures by default.

### Description of development approach
Scanning document by paragraph in the desired direction and analyzing styles within each paragraph.
### Testing strategy:
Tested manually in Chrome and Firefox.
Added a system test.
### Known issues with pull request:
* In addition to same style navigation I also added different style navigation command because it's implementation turned out to be very similar - I hope that's ok.
* This PR will have conflicting changes with #16031 - I will resolve conflicts whenever either of these PRs is merged.
* ~~Another quirk to be aware of: in this PR I treat paragraphs separately. For example if we have an entire document written in the same style, then jump to next same style command would iterate over paragraphs. In theory we can try to merge paragraphs and parts of paragraphs written in the same style together, but this would make code more complicated and I personally feel current approach is more intuitive, but LMK if you think otherwise.~~ Update: as requested, now I changed the code so that every chunk found by this pR can span multiple paragraphs.
### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

mltony committed:
Text style navigation

mltony commented:
The bot says some system test with tag Installer is failing. I couldn't find any information on waht test is failing. Not sure if this is flaky test. If it's not flaky - could someone help me to find which test is failing? It wasn't very clear to me based on bot's comment.

seanbudd commented:
Yes I think that test failure is unrelated

LeonarddeR commented on source/browseMode.py:
```diff
@@ -2001,6 +2040,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
+						)
+					):
+						# Found text that matches desired style!
+						startOffset = sum([
```
You are assuming here that the number of characters in a textInfo is equal to the number of characters in a Python string, right?
This is not true, as many Offset based textInfo implementations including virtual buffers work with UTF16 characters. There, emoji are treated as two characters, whereas they are one character in Python.
You should be able to convert between widestring and Python offsets with the tools in the textUtils module.

CyrilleB79 commented on source/browseMode.py:
```diff
@@ -2001,6 +2040,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
+			result = initialTextInfo.move(textInfos.UNIT_CHARACTER, -1, endPoint='start')
+			if result == 0:
+				# Translators: Error message for same/different style QuickNav command
+				ui.message(_("Cannot determine current style"))
```
Is it an expected error? If yes, could you describe a case?
* If the error is expected in some case, there is no point in raising an error. Just return if needed, and log a debugWarning if it seems of interest, so that we can see that we are in a corner case.
* If the error is not expected: there is no point in using a `ui.message`. The user will already have no message, when expecting one, and there will be an error in the log.

mltony replied:
If I simply return from this generator, then `_quickNavScript` function that calls the generator would report "no next same style text" - which is wrong.
The error is not really expected, however I think this is a good style to check for potential problems in code. Like there are many different TextInfo implementations and all of them have slightly different behaviors. So it doesn't hurt to speak a message if one of sanity checks doesn't pass. However if you strongly feel that we should raise Exception without a message - I can remove it.

CyrilleB79 replied:
OK. Thanks for the clarification.

This pattern with a UI message and an error is not so common in NVDA. But if NV Access is OK, I have no objection.

Generally It may have been still better to call the `ui.message` in the calling function in a try block. But in the present case, it does not seem applicable due to code being factorized.

mltony committed:
Addressing comment by @LeonarddeR

CyrilleB79 commented on source/browseMode.py:
```diff
@@ -2001,6 +2041,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
+		if result == 0:
+			result = initialTextInfo.move(textInfos.UNIT_CHARACTER, -1, endPoint='start')
+			if result == 0:
+				# Translators: Error message for same/different style QuickNav command
```
```suggestion
				# Translators: Error message for same/different style quick navigation command
```

CyrilleB79 commented on source/browseMode.py:
```diff
@@ -2001,6 +2041,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
+			or not isinstance(styles[0], textInfos.FieldCommand)
+			or styles[0].command != 'formatChange'
+		):
+			# Translators: Error message for same/different style QuickNav command
```
```suggestion
			# Translators: Error message for same/different style quick navigation commands
```

CyrilleB79 commented on user_docs/en/changes.t2t:
```diff
@@ -7,6 +7,7 @@ What's New in NVDA
 
 == New Features ==
 - In Windows 11, NVDA will announce alerts from voice typing and suggested actions including the top suggestion when copying data such as phone numbers to the clipboard (Windows 11 2022 Update and later). (#16009, @josephsl)
+- Added same style and different style navigation QuickNav commands, not assigned to any gestures.
```
1. quick nav -> quick navigation
2. Add the issue number
3. Added credit to you; I think that you have added it in the other PRs, so I guess you agree also for this one.

```suggestion
- Added same style and different style quick navigation commands, not assigned to any gestures. (#16000, @mltony)
```

CyrilleB79 reviewed (commented):
This feature is really promising.

I have only a few small GUI / documentation changes to suggest.

I have not tested the code though.

CyrilleB79 commented:
In an Outlook message, I was hoping to jump to "answers in red" sent by a colleague of mine.

Unfortunately, it does not work. I get errors such as the following when calling all 4 commands (here with Move to next different style text command):

```
ERROR - scriptHandler.executeScript (16:39:49.269) - MainThread (9984):
error executing script: <bound method BrowseModeTreeInterceptor.addQuickNav.<locals>.<lambda> of <nvdaBuiltin.appModules.outlook.MailViewerTreeInterceptor object at 0x01DBFF90>> with gesture 'j'
Traceback (most recent call last):
  File "scriptHandler.py", line 295, in executeScript
    script(gesture)
  File "browseMode.py", line 508, in <lambda>
    script = lambda self,gesture: self._quickNavScript(gesture, itemType, "next", nextError, readUnit)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "browseMode.py", line 466, in _quickNavScript
    item = next(iterFactory(direction, info))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "browseMode.py", line 2111, in _iterTextStyle
    styles = self._extractStyles(initialTextInfo)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "browseMode.py", line 2075, in _extractStyles
    if field.field['role'] == controlTypes.Role.MARKED_CONTENT:
       ~~~~~~~~~~~^^^^^^^^
KeyError: 'role'
```

CyrilleB79 commented:
Regarding unassigned keys:
IMO, if this feature works as advertised, it would be much more interesting to have default assigned keys for it and keep QuickNav Text paragraph navigation commands unassigned (#16031).

Maybe we should merge this PR and 16031 with unassigned gestures and test them during alpha stage. Then at the end of alpha stage, we can figure if one or both of these feature need to have an assigned gesture.

mltony committed:
Update source/browseMode.py

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

mltony committed:
Update source/browseMode.py

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

mltony committed:
Update user_docs/en/changes.t2t

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

mltony committed:
Fix outlook exception

lukaszgo1 commented on source/browseMode.py:
```diff
@@ -2001,6 +2041,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
+	STYLE_ATTRIBUTES = frozenset([
+		'background-color',
+		'color',
+		'font-family,',
```
Shouldn't this be:
```suggestion
		'font-family',
```

mltony committed:
Update source/browseMode.py

Co-authored-by: Łukasz Golonka <lukasz.golonka@mailbox.org>

seanbudd commented on source/browseMode.py:
```diff
@@ -439,10 +441,24 @@ def _iterNodesByType(self,itemType,direction="next",pos=None):
+	def _iterTextStyle(
+			self,
+			kind: str,
+			direction: str = "next",
```
this could probably use `documentBase._Movement`

seanbudd commented on source/browseMode.py:
```diff
@@ -439,10 +441,24 @@ def _iterNodesByType(self,itemType,direction="next",pos=None):
 			iterFactory=self._iterNotLinkBlock
+		elif itemType in ["sameStyle", "differentStyle"]:
+			def iterFactory(
+					direction: str,
```
this could probably use `documentBase._Movement`

seanbudd commented on source/browseMode.py:
```diff
@@ -439,10 +441,24 @@ def _iterNodesByType(self,itemType,direction="next",pos=None):
+			self,
+			kind: str,
+			direction: str = "next",
+			pos: textInfos.TextInfo = None,
```
```suggestion
			pos: textInfos.TextInfo | None = None,
```

seanbudd commented on source/browseMode.py:
```diff
@@ -439,10 +441,24 @@ def _iterNodesByType(self,itemType,direction="next",pos=None):
+		elif itemType in ["sameStyle", "differentStyle"]:
+			def iterFactory(
+					direction: str,
+					info: textInfos.TextInfo,
```
```suggestion
					info: textInfos.TextInfo | None,
```

seanbudd commented on source/browseMode.py:
```diff
@@ -2001,6 +2041,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
 			item1=item2
 
+	STYLE_ATTRIBUTES = frozenset([
+		'background-color',
```
I know this is a pain, but could you please use double quotes consistently rather than mixed double and single quotes? NVDA uses double quotes by default

seanbudd commented on source/browseMode.py:
```diff
@@ -2001,6 +2041,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
+	) -> "textInfos.TextInfo.TextWithFieldsT":
+		"""
+		This function calls TextInfo.getTextWithFields(), and then processes fields in the following way:
+		1. Highlighted(marked) text is currently reported as Role.MARKED_CONTENT, and not formatChange.
```
```suggestion
		1. Highlighted (marked) text is currently reported as Role.MARKED_CONTENT, and not formatChange.
```

seanbudd commented on source/browseMode.py:
```diff
@@ -2001,6 +2041,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
+		and set its value according to presence of Role.MARKED_CONTENT.
+		2. Then we drop all control fields, leaving only formatChange fields and text.
+		"""
+		stack = [{}]
+		result = []
```
can these get type info added. It's useful for objects like to know what child item types to expect

seanbudd commented on source/browseMode.py:
```diff
@@ -2001,6 +2041,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
+			elif isinstance(field, str):
+				result.append(field)
+			else:
+				raise RuntimeError
```
can you note that this raises a runtime error in the docstring?
Please add a relevant message to the runtime error

seanbudd commented on source/browseMode.py:
```diff
@@ -2001,6 +2041,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
+					}
+					result.append(field)
+				else:
+					raise RuntimeError()
```
Please add a relevant message to the runtime error

seanbudd commented on source/browseMode.py:
```diff
@@ -2001,6 +2041,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
+	def _iterTextStyle(
+			self,
+			kind: str,
+			direction: str = "next",
```
this could probably use `documentBase._Movement`

seanbudd commented on source/browseMode.py:
```diff
@@ -2001,6 +2041,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
+			self,
+			kind: str,
+			direction: str = "next",
+			pos: textInfos.TextInfo = None
```
```suggestion
			pos: textInfos.TextInfo | None = None
```

seanbudd commented on source/browseMode.py:
```diff
@@ -2001,6 +2041,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
+		):
+			# Translators: Error message for same/different style quick navigation commands
+			ui.message(_("Cannot determine current style"))
+			raise RuntimeError
```
Please add a relevant message to the runtime error

seanbudd commented on source/browseMode.py:
```diff
@@ -2001,6 +2041,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
+		tmpInfo = pos.copy()
+		tmpInfo.expand(textInfos.UNIT_PARAGRAPH)
+		paragraph.setEndPoint(tmpInfo, which='endToEnd' if direction == 'next' else 'startToStart')
+		while True:
```
I'm concerned about being caught in `while True` 

```suggestion
		MAX_ITER_LIMIT = 10**6
		for _ in range(MAX_ITER_LIMIT):
```

seanbudd commented on tests/system/robot/chromeTests.py:
```diff
@@ -2473,3 +2473,89 @@ def test_i13307():
+
+def test_styleNav():
+	""" Tests that same style and different style navigation work correctly in browse mode.
+	Bydefault these commands don't have assigned gestures,
```
```suggestion
	By default these commands don't have assigned gestures,
```

mltony committed:
Update source/browseMode.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update source/browseMode.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update source/browseMode.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update source/browseMode.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update tests/system/robot/chromeTests.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Addressing comments by @seanbudd

mltony committed:
Addressing comments by @seanbudd

mltony committed:
lint

mltony committed:
Merge branch 'master' into styleNav

mltony committed:
lint

seanbudd commented:
Hi @mltony , There are 5 review comments from last review unresolved. I think they were missed as they were automatically hidden due to the many review comments.

mltony committed:
Update source/browseMode.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Addressing comments by @seanbudd

mltony committed:
Merge branch 'master' into styleNav

mltony commented:
@seanbudd my bad, didn't see those comments. Now I addressed all of them.

mltony committed:
Fixing more quotes

mltony committed:
lint

CyrilleB79 commented:
Hi Tony,

The feature does not work at all in Word (browse mode):
With Word legacy (object model):
* move to next same style text jumps to next word, no matter the formatting.
* move to next different style text causes an error:
```
ERROR - scriptHandler.executeScript (22:58:32.703) - MainThread (8308):
error executing script: <bound method BrowseModeTreeInterceptor.addQuickNav.<locals>.<lambda> of <NVDAObjects.window.winword.WordDocumentTreeInterceptor object at 0x00C6C370>> with gesture '<'
Traceback (most recent call last):
  File "scriptHandler.py", line 295, in executeScript
    script(gesture)
  File "browseMode.py", line 559, in <lambda>
    script = lambda self,gesture: self._quickNavScript(gesture, itemType, "next", nextError, readUnit)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "browseMode.py", line 517, in _quickNavScript
    item = next(iterFactory(direction, info))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "browseMode.py", line 2235, in _iterTextStyle
    paragraph.collapse(end=True)
  File "treeInterceptorHandler.py", line 234, in collapse
    return self.innerTextInfo.collapse(end=end)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "NVDAObjects\window\winword.py", line 1048, in collapse
    raise RuntimeError

```

If instead you use UIA in Word:
* move to next same style text jumps to next different style text instead
* move to next different style text reports "No next different style text", even when there are various formatting.

Could you double check all the Word/Outlook use cases? Thanks.

mltony committed:
Fix for MSWord

mltony commented:
@CyrilleB79 , I investigated a bit into MSWord. Pushed one more commit to address one issue with MSWord - now I am merging identical styles together. This is especially important for MSWord since textInfos there cchange their format every word - but - they mark all the words with a detected language, while  spaces in between words have no assigned language. Since we filter out language from the list of styles, the remaining styles are identical and should be merged.
Also I filed two issues - these appear to be problems with MSWOrd-specific textInfo implementation: #16171 and #16172.

mltony committed:
lint

mltony committed:
Merge branch 'master' into styleNav

mltony committed:
minor

CyrilleB79 commented:
@mltony the feature does not seem to work with Word legacy; I did not re-test with UIA.

Could you double-check that all command types work in Word with and without UIA? Let me know when you have re-tested this successfully. Thanks.

mltony commented:
@CyrilleB79  - as I mentioned #16171 and #16172 are word specific problems that prevent style navigation to work correctly in MSWord. Both issues affect both legacy and UIA word support. Until both are fixed, style navigation won't work correctly in MSWord.

CyrilleB79 commented:
@mltony, sorry to come a bit late with it, but I had not paid attention to the following known issue:
> Another quirk to be aware of: in this PR I treat paragraphs separately. For example if we have an entire document written in the same style, then jump to next same style command would iterate over paragraphs. In theory we can try to merge paragraphs and parts of paragraphs written in the same style together, but this would make code more complicated and I personally feel current approach is more intuitive, but LMK if you think otherwise.wouldn't seem very logic

My personal opinion is that merging same style across paragraph boundaries is more than desirable. For me, the main use case of this PR is : "See my comments in red the message below". In a message with many bullet paragraphs (e.g. meeting minutes), with only few comments in red, I do not expect the jump commands to jump to each bullet.

CyrilleB79 commented:
> @CyrilleB79 - as I mentioned #16171 and #16172 are word specific problems that prevent style navigation to work correctly in MSWord. Both issues affect both legacy and UIA word support. Until both are fixed, style navigation won't work correctly in MSWord.

Re #16171: as mentioned there, word is able to read text with formatting information if NVDA is configured for this. Thus you should probably use the same technique for style navigation.

Re #16172: I have not reviewed the code. But it seems strange to me that you cannot work it around, e.g. catching the exception to indicate the end of the document. Actually Word has a browse mode with a working quick nav capability. Why is style navigation impacted and other quick nav commands not?

I hope you can fix #16171 and #16172 in this PR. Especially because Outlook uses Word browse mode on by default, and because the "See my comments in red in the message below" is a very common use case.

Though, if you confirm that you cannot fix them, you should:
* mention it in known issue section
* disable these commands in Word/Outlook so that we do not expect them to work

XLTechie commented:
What @CyrilleB79 describes is the behavior I would expect also. I'm not sure how 
this feature is useful if it doesn't act that way.

mltony committed:
Addressing comments

mltony commented:
@CyrilleB79, @XLTechie,
I addressed your comments. Now same style chunks can span across multiple paragraphs. This made the code significantly more complicated as I warned before. Took me quite some time to debug that.
Also I made this PR compatible with MS Word. I figured out how to retrieve formatting information there. Also I still think #16172 is a bug - as it shouldn't throw an error. But for now I am catching this error in my code, so this PR works as expected and is not blocked on #16172.

CyrilleB79 commented:
Thanks @mltony. This actually works much better in Word. 

There still seems to be little issues with links of e-mails and URLs. Try to type the following:
My e-mail is [foo.bar@example.com](mailto:foo.bar@example.com) and I’ll provide a link on https://github.com/nvaccess/nvda/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc. Is it OK.

Depending on where you have initially put the cursor, you get unexpected results. E.g. cursor before the link, or after its first chararacter, or in the middle, or on the penultimate character, or on the ultimate character, or just after the last character of the link, etc.
I wonder if it's related to punctuation symbols or to some word strange behaviour.

mltony committed:
Addressing comments

mltony commented:
@CyrilleB79 ,
I debugged this and drilled it down to #16196 which happnes to be a duplicate of #11427, which according to Sean is blocked on an external fix.
It seems that the issue is only applicable to links in MSWord. So I worked around this by ignoring color of links in MSWord.

mltony committed:
Minor

mltony committed:
Merge branch 'master' into styleNav

mltony committed:
Switching to moveToCodepointOffset

mltony commented:
I updated this to use moveToCodepointOffset function - this is now ready for review. I also fixed a small bug in moveToCodepointOffsets and added more unit tests to cover that bug.

seanbudd commented on tests/unit/test_textInfos.py:
```diff
@@ -221,11 +221,25 @@ class TestMoveToCodepointOffsetInOffsetsTextInfo(unittest.TestCase):
 		info = obj.makeTextInfo(Offsets(0, 0))
-		info.expand(textInfos.UNIT_STORY)
+		info._startOffset = info._endOffset = prefixOffset
+		print(info._getStoryLength())
```
please use logging not print statements

seanbudd reviewed (commented):
Changes generally look good to me

Qchristensen reviewed (approved):
Simple change for the user guide etc, reads well.

mltony committed:
Update source/browseMode.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Addressing comments

seanbudd reviewed (approved):
Thanks @mltony
